tests/integration/__init__.py

```diff
@@ -8,16 +9,116 @@ class IntegrationTestBase(TestBase):
     Base class for express integration tests.
     """
 
-    def setUp(self):
-        super(IntegrationTestBase, self).setUp()
+    def work_dir(self, version: str, subdir: str):
+        return os.path.join(
+            self.rootDir, "fixtures", self.application, version, subdir
+        )
 
-    def tearDown(self):
-        super(IntegrationTestBase, self).setUp()
+    def get_reference_value(self):
+        """Obtain the appropriate reference value to use as an expected value for the
+        parser extraction."""
+        sub_path = self.subdir.split("/")[-1]
+        expected = self.references[self.version][sub_path][self.property]
+        reference_index = self.test_config.pop("reference_index", None)
```

Comment:
Oh hey, where do we define `self.subdir` or `self.version`, `self.property`, and `self.test-config` (or `self.application` in the `__init__`)? Does this happen in the `setattr` call in the `create_test` method?

If they're expected to be subclassed or derived from the yaml, do you think it'd be worth just raising `NotImplementedError` in the base class if we try to use them before they're set?

Reply:
Right, in an intermediate step I had these attributes defined in an `__init__` but it was causing unintended side effects with other tests inheriting from `TestIntegrationBase`. I will isolate the class hierarchy and be more explicit with the expected attributes.

Reply:
Ok this way attributes are at least specified in the `__init__` but still get set dynamically because of the nature of the test class lifecycle.